docs/maintaining.md

```diff
@@ -92,3 +92,118 @@ made to always install the latest released version.
+
+## `grafana/prometheus` Maintainence
+
+Grafana Labs is using the Agent for their internal monitoring and want to take
+advantage of the Agent to proof-of-concept additions to Prometheus before they
+get moved upstream. A `grafana/prometheus` repository is maintained by Grafana
+Labs where all non-trivial changes will go first. Doing this allows for getting
+Cortex-specific changes moving along faster and providing strong evidence
+towards its usefulness and correctness when it becomes part of an upstream PR.
+
+We are commiting ourselves to doing the following:
+
+1. Always use a recent Prometheus release: the Agent will always vendor a
+   recent Prometheus release and not Prometheus master. We want the Agent's
+   Prometeheus roots to be stable.
+2. Keep changes mergeable upstream: we want to continue to be good OSS citizens,
+   and we intend that all features we add to our Prometheus repository will
+   become an upstream PR. We will maintain our repository in a way that supports
+   doing this.
+3. Reduce code drift: The code the Agent uses on top of Prometheus will be
+   layered on top of a Prometheus release rather than sandwiched in between.
+
+Maintainence of the `grafana/prometheus` repository revolves around feature
+branches (named `feat-SOME-FEATURE`) and release branches (named
+`release-vX.Y.Z-grafana`). The release branches will always use the same release
+version as the `prometheus/prometheus` release it is based off of.
+
+By adding features to the `grafana/prometheus` repository first, we are
+committing ourselves to extra maintenance of features that have not yet been
+merged upstream. Feature authors will have to babysit their features to
+coordinate with the Prometheus release schedule to always be compatible. One the
+feature is merged upstream, this burden of maintaining is made easier as the
+Prometheus team can more easily sync breaking changes across the whole
+repository.
+
+We are purposefully carrying this extra burden because we intend to ultimately
+make Prometheus better and contribute all of our enhancements upstream. We want
+to strive to benefit the Prometheus ecosystem at large.
+
+### Creating a New Feature
+
+For `grafana/prometheus` maintainers to create a new feature, they will do the
+following:
+
+1. Create a feature branch in `grafana/prometheus` based on the latest release
+   tag that `grafana/prometheus` currently has a release branch for. The feature
+   branch should follow the naming convention `feat-<feature name>`.
+2. Implement the feature and open a PR to merge the feature branch into the
+   associated `grafana/prometheus` release branch.
+3. Once the release branch is updated, open a PR to update `grafana/agent` to
+   use the latest release branch SHA.
+
+### Updating an Existing Feature
+
+If a feature branch that was already merged to a release branch needs to be
+updated for any reason:
+
+1. Push directly to the feature branch or open a PR to merge changes into that
```

Comment:
I'm not sure about encouraging trunk-based development by pushing directly. Do you feel confident about this?

Reply:
I do, because review of changes that were pushed directly to the feature branch can still be made in the PR which merges it into the release branch.

Reply:
Sounds good. My worry here was about the new changes you would push (if needed) - out of the two I would almost always prefer to open the PR unless it's an emergency and then I wouldn't even bother following _any_ protocol.